From 02465e72ee246d39bc6a29935c63faab47888cb0 Mon Sep 17 00:00:00 2001
From: Antonio Borneo <antonio.borneo@foss.st.com>
Date: Wed, 12 Jul 2023 10:37:55 +0200
Subject: [PATCH 324/841] feat(stm32mp2): add watchdog support

Add the private functions to read OTP and update shadow (without
locking them).
Initialize the watchdog driver and refresh it after the security
setup.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git b491f8162eb09998dfdd4a66596c7cbfe2d8cdca feat(stm32mp2): add watchdog support]
Signed-off-by: Antonio Borneo <antonio.borneo@foss.st.com>
Change-Id: I20c08950b838db795b9f905708c7fce7162c5256
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/317862
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 plat/st/stm32mp2/bl2_plat_setup.c   |  7 +++
 plat/st/stm32mp2/stm32mp2_private.c | 81 +++++++++++++++++++++++++++++
 2 files changed, 88 insertions(+)

diff --git a/plat/st/stm32mp2/bl2_plat_setup.c b/plat/st/stm32mp2/bl2_plat_setup.c
index 718f21764..f4bdfc8e7 100644
--- a/plat/st/stm32mp2/bl2_plat_setup.c
+++ b/plat/st/stm32mp2/bl2_plat_setup.c
@@ -20,6 +20,7 @@
 #include <drivers/st/regulator.h>
 #include <drivers/st/regulator_fixed.h>
 #include <drivers/st/stm32_console.h>
+#include <drivers/st/stm32_iwdg.h>
 #include <drivers/st/stm32_rng.h>
 #include <drivers/st/stm32mp_pmic2.h>
 #include <drivers/st/stm32mp_reset.h>
@@ -229,6 +230,12 @@ void bl2_el3_plat_arch_setup(void)
 		panic();
 	}
 
+	if (stm32_iwdg_init() < 0) {
+		panic();
+	}
+
+	stm32_iwdg_refresh();
+
 	stm32_save_boot_info(boot_context);
 
 	if (stm32mp_uart_console_setup() != 0) {
diff --git a/plat/st/stm32mp2/stm32mp2_private.c b/plat/st/stm32mp2/stm32mp2_private.c
index 6bf3c7007..f1e76d360 100644
--- a/plat/st/stm32mp2/stm32mp2_private.c
+++ b/plat/st/stm32mp2/stm32mp2_private.c
@@ -9,6 +9,7 @@
 
 #include <arch_helpers.h>
 #include <drivers/arm/gicv2.h>
+#include <drivers/st/stm32_iwdg.h>
 #include <lib/mmio.h>
 #include <lib/xlat_tables/xlat_tables_v2.h>
 #include <plat/common/platform.h>
@@ -416,6 +417,86 @@ bool stm32mp_is_auth_supported(void)
 	return supported;
 }
 
+uint32_t stm32_iwdg_get_instance(uintptr_t base)
+{
+	uint32_t instance = UINT32_MAX;
+
+	switch (base) {
+	case IWDG1_BASE:
+		instance = IWDG1_INST;
+		break;
+	case IWDG2_BASE:
+		instance = IWDG2_INST;
+		break;
+	default:
+		break;
+	}
+
+	if (instance == UINT32_MAX) {
+		panic();
+	}
+
+	return instance;
+}
+
+uint32_t stm32_iwdg_get_otp_config(uint32_t iwdg_inst)
+{
+	uint32_t iwdg_cfg = 0U;
+	uint32_t otp_value;
+
+	if (stm32_get_otp_value(HCONF1_OTP, &otp_value) != 0U) {
+		panic();
+	}
+
+	if ((otp_value & BIT(iwdg_inst + HCONF1_OTP_IWDG_HW_POS)) != 0U) {
+		iwdg_cfg |= IWDG_HW_ENABLED;
+	}
+
+	if ((otp_value & BIT(iwdg_inst + HCONF1_OTP_IWDG_FZ_STOP_POS)) != 0U) {
+		iwdg_cfg |= IWDG_DISABLE_ON_STOP;
+	}
+
+	if ((otp_value & BIT(iwdg_inst + HCONF1_OTP_IWDG_FZ_STANDBY_POS)) != 0U) {
+		iwdg_cfg |= IWDG_DISABLE_ON_STANDBY;
+	}
+
+	return iwdg_cfg;
+}
+
+#if defined(IMAGE_BL2)
+uint32_t stm32_iwdg_shadow_update(uint32_t iwdg_inst, uint32_t flags)
+{
+	uint32_t otp_value;
+	uint32_t otp_idx;
+	uint32_t result;
+
+	if (stm32_get_otp_index(HCONF1_OTP, &otp_idx, NULL) != 0U) {
+		panic();
+	}
+
+	if (stm32_get_otp_value_from_idx(otp_idx, &otp_value) != 0U) {
+		panic();
+	}
+
+	if ((flags & IWDG_DISABLE_ON_STOP) != 0U) {
+		otp_value |= BIT(iwdg_inst + HCONF1_OTP_IWDG_FZ_STOP_POS);
+	}
+
+	if ((flags & IWDG_DISABLE_ON_STANDBY) != 0U) {
+		otp_value |= BIT(iwdg_inst + HCONF1_OTP_IWDG_FZ_STANDBY_POS);
+	}
+
+	result = bsec_write_otp(otp_value, otp_idx);
+	if (result != BSEC_OK) {
+		return result;
+	}
+
+	/* Do not sticky lock OTP_IWDG (read and write) */
+
+	return BSEC_OK;
+}
+#endif
+
 bool stm32mp_skip_boot_device_after_standby(void)
 {
 	return false;
-- 
2.47.2

